backend/core/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisService.java

```diff
@@ -340,4 +355,81 @@ private static Predicate<? super SonarLintRuleDefinition> isIncludedByConfigurat
       return false;
     };
   }
+
+  @EventListener
+  public void onPluginsSynchronized(PluginsSynchronizedEvent event) {
+    checkIfReadyForAnalysis(configurationRepository.getBoundScopesToConnection(event.getConnectionId())
+      .stream().map(BoundScope::getConfigScopeId).collect(Collectors.toSet()));
+  }
+
+  @EventListener
+  public void onConfigurationScopeAdded(ConfigurationScopesAddedEvent event) {
+    checkIfReadyForAnalysis(event.getAddedConfigurationScopeIds());
+  }
+
+  @EventListener
+  public void onConfigurationScopeRemoved(ConfigurationScopeRemovedEvent event) {
+    analysisReadinessByConfigScopeId.remove(event.getRemovedConfigurationScopeId());
+  }
+
+  @EventListener
+  public void onBindingConfigurationChanged(BindingConfigChangedEvent event) {
+    var configScopeId = event.getConfigScopeId();
+    analysisReadinessByConfigScopeId.remove(configScopeId);
+    if (!checkIfReadyForAnalysis(configScopeId)) {
+      client.didChangeAnalysisReadiness(new DidChangeAnalysisReadinessParams(Set.of(configScopeId), false));
+    }
+  }
+
+  @EventListener
+  public void onAnalyzerConfigurationSynchronized(AnalyzerConfigurationSynchronized event) {
+    checkIfReadyForAnalysis(event.getConfigScopeIds());
+  }
+
+  @EventListener
+  public void onAnalyzerConfigurationSynchronized(ConfigurationScopesSynchronizedEvent event) {
+    checkIfReadyForAnalysis(event.getConfigScopeIds());
+  }
+
+  private void checkIfReadyForAnalysis(Set<String> configurationScopeIds) {
+    var readyConfigScopeIds = configurationScopeIds.stream()
+      .filter(this::isReadyForAnalysis)
+      .collect(toSet());
+    saveAndNotifyReadyForAnalysis(readyConfigScopeIds);
+  }
+
+  private boolean checkIfReadyForAnalysis(String configurationScopeId) {
+    if (isReadyForAnalysis(configurationScopeId)) {
```

Comment:
This is not atomic if I am not mistaken

Reply:
These methods are invoked from event listeners, which are called synchronously and serially, so they are not accessed concurrently. We make this assumption in other places too. I agree later if we change the events delivery it might break, but I don't want to over-complexify the code. Do you have a suggestion?